### DISCUSSION TITLE
Clarification on the "Implementation Approach" column

### DISCUSSION THREAD

**Discussion by Banashek** [TechEmpower/FrameworkBenchmarks#7483] [General]
Hi :wave: ,

I was looking through a handful of implementations and wanted to ask about the spirit of the "Implementation Approach" column in the data-grid.

The description is thus:

![image](https://user-images.githubusercontent.com/7094571/180263245-bd5bf44c-ab7d-4dd8-a518-3fb80c2d7a1a.png)

The description is interesting because it is ever so slightly different from a mental model that I have, which paints a dichotomy between:

1) Framework used as advertised
2) Framework used in a non-standard (potentially non-realistic) way, eschewing features and manually re-implementing parts of the framework for higher performance.

Perhaps I am conflating the "Classification" field with the "Implementation Approach" field and that is the entirety of my confusion, but want to elaborate on my thoughts below to clarify:

---

I'm going to pick on aspnet here because I like the framework and enjoy using it, but it is not the only implementation which I think is not-realistic-marked-as-realistic:

![image](https://user-images.githubusercontent.com/7094571/180264104-5a3afc83-1983-4bb7-adf0-efe6cce894c9.png)

As we can see, all of the aspnet core fortune tests are marked as realistic.

However, drilling down a little bit we can see that the implementation for the base fortunes benchmark is [writing to a buffered writer (essentially stringbuilding)](https://github.com/TechEmpower/FrameworkBenchmarks/blob/ed226ae0c87831a9b85f0d4bd2e78329cec4c37a/frameworks/CSharp/aspnetcore/PlatformBenchmarks/BenchmarkApplication.Fortunes.cs#L28) which [is also defined in the project](https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/CSharp/aspnetcore/PlatformBenchmarks/BufferWriter.cs)

---

This could very well be within the guidelines of "Realistic Approach", but to me seems a bit disingenuous.
The network effects are non-zero, and in some cases [the data is used in official material](https://devblogs.microsoft.com/dotnet/announcing-asp-net-core-in-net-5/#performance-improvements).

I do understand it's difficult to fairly represent the spectrum of frameworks (a "realistic" approach for a barebones framework may be the same code as the c# above), but was curious if there were any thoughts of a more clear delineation.

---

Thanks for expending the time/$ to run these, it's always interesting to see the shape of the data and how it evolves! :bow:

**Comment by bhauer** [comment]
The subject you raise is at least adjacent to some topics we've discussed in the past. Those past discussions raised some interesting ideas, but changes have not yet materialized.

We have considered, for example, asking framework maintainers to "endorse" certain implementations. An endorsed implementation could be represented by another badge similar to the "T" icon on frameworks that are included in the hardware performance calculation. Endorsement is effectively a framework maintainer conveying the implementation is consistent with how they expect the framework to be used in production environments.

Similarly, we considered adding more types of "Implementation Approach." You can imagine, for example, a range of values such as the following, in roughly descending order of perceived value: "Endorsed/Canonical," "Production-Ready," "Experimental," and "Stripped."

Related, past discussions included an intent to allow maintainers to document and describe the approaches they've taken and have that then made visible on the results web site, along with links to the frameworks, and so on. Basically give the contributors a way to communicate about their implementation approach and weight in on how readers can interpret the results.

**Reply by Banashek** [reply]
Thanks for reiterating past talking points, I must not have looked hard enough for these previous posts.
The ideas you present seem well thought out.
I'll be interested to see how the project continues to evolve!
